### ISSUE TITLE
Windows launcher finds but fails to launch Python <3.9

### ISSUE THREAD

**Issue by kroeschl** [python/cpython#93263]
**Bug report**

The `py` launcher finds but will not run Python versions before at latest 3.9.12. Found on a system with Python 3.6,7,8,9,10 and 2.7 installed. Example trying to launch Python 3.6:
```
> py -3.6 -V
Installed Pythons found by C:\Windows\py.exe Launcher for Windows *

Requested Python version (3.6) not installed, use -0 for available pythons
Python 3.6 not found!

 -3.10-64
 -3.9-64
 -3.8.10-64
 -3.7.9-64
 -3.6.8-64
 -2.7.18-64
```

<details>
<summary>Output with debugging enabled</summary>

```
> PYLAUNCH_DEBUG=1 py -3.6 -V
launcher build: 32bit
launcher executable: Console
File 'C:\Users\Administrator\AppData\Local\py.ini' non-existent
File 'C:\Windows\py.ini' non-existent
Called with command line: -3.6 -V
locating Pythons in 64bit registry
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\2.7.18\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.6.8\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.7.9\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.8.10\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe is a 64bit executable
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: : The system cannot find the path specified.
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: unable to open PythonCore key in HKLM
locating Pythons in native registry
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\2.7.18\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\2.7.18\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\2.7.18\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\2.7.18\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.6.8\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.6.8\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.6.8\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.6.8\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.7.9\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.7.9\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.7.9\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.7.9\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.8.10\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.8.10\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.8.10\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.8.10\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.12\python.exe: already found
locate_pythons_for_key: unable to open PythonCore key in HKLM
locate_pythons_for_key: unable to open PythonCore key in HKLM
search for Python version '3.6' found no interpreter
found no configured value for 'python'
search for default Python found version 3.10 at 'C:\Users\Administrator\.pyenv\pyenv-win\versions\3.10.4\python.exe'
Installed Pythons found by C:\Windows\py.exe Launcher for Windows *

Requested Python version (3.6) not installed, use -0 for available pythons
Python 3.6 not found!

 -3.10-64
 -3.9-64
 -3.8.10-64
 -3.7.9-64
 -3.6.8-64
 -2.7.18-64
```
</details>

<details>
<summary>Registry tree</summary>

```
>reg query HKCU\SOFTWARE\Python\PythonCore

HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore
    DisplayName    REG_SZ    Python Software Foundation
    SupportUrl    REG_SZ    https://www.python.org/

HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\2.7.18
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.10
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.10.4
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.6.8
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.7.9
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.8.10
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.9
HKEY_CURRENT_USER\SOFTWARE\Python\PythonCore\3.9.12
```
</details>

From poking through the launcher source, it looks like this is caused by:
- [`validate_version()`](https://github.com/python/cpython/blob/949dbf97ba09da3cfb243d5ad7f90967ad15f354/PC/launcher.c#L1198) only allows specifying up to the minor version, not patch.
- `find_python_by_version()` requires that the version either be single-digit (e.g. `py -3`) or an exact match with whatever we have in our list of found Python versions.
- Our registry-based Python lookup gets the [version from the branch names](https://github.com/python/cpython/blob/949dbf97ba09da3cfb243d5ad7f90967ad15f354/PC/launcher.c#L283) under `..\SOFTWARE\Python\PythonCore`, which defaults to only the full version for Python ending somewhere between 3.8.10 and 3.9.12.

The current launcher should probably be compatible with at least Python versions still in support. My current workaround for this issue is to `reg copy` the fully-qualified branches under `PythonCore` to branches only specified to the minor version:
```
reg copy HKCU\SOFTWARE\Python\PythonCore\3.8.10 HKCU\SOFTWARE\Python\PythonCore\3.8 /s /f
```

**Your environment**

- CPython versions tested on: 3.10.4
  Also installed on the system: 2.7.18, 3.6.8, 3.7.9, 3.8.10, 3.9.12
- Operating system and architecture: Windows 10 64-bit

**Comment by eryksun** [comment]
The standard Python installer names the subkey of "PythonCore" as "major.minor[-32]". Using "major.minor.micro[-32]" is not supported.

**Comment by kroeschl** [comment]
It turns out [`pyenv-win`](https://github.com/pyenv-win/pyenv-win) was changing the registry entries in order to allow side-by-side installation of different micro versions. I thought I had verified that it wasn't doing anything weird to the registry entries, but I must have messed up something in my testing.

If I install multiple Python versions via the standard installers, everything works fine. Closing this out--sorry for the noise.
